pkg/gds/models/v1/models.go

```diff
@@ -45,6 +45,51 @@ func SetAdminVerificationToken(vasp *pb.VASP, token string) (err error) {
 	return nil
 }
 
+// GetVerificationChangeLog from the extra data on the VASP record.
+func GetVerificationChangeLog(vasp *pb.VASP) (_ []*VerificationChangeLogEntry, err error) {
+	// If the extra data is nil, return nil (no change log).
+	if vasp.Extra == nil {
+		return nil, nil
+	}
+
+	// Unmarshal the extra data field on the VASP.
+	extra := &GDSExtraData{}
+	if err = vasp.Extra.UnmarshalTo(extra); err != nil {
+		return nil, err
+	}
+	return extra.GetVerificationChangeLog(), nil
+}
+
+// Append a VerificationChangeLogEntry to the extra data on the VASP record.
+func AppendVerificationChangeLog(vasp *pb.VASP, entry *VerificationChangeLogEntry) (err error) {
+	// Validate current state.
+	if entry.CurrentState > uint32(pb.VerificationState_ERRORED) {
+		return fmt.Errorf("cannot set verification state to unsupported value %d", entry.CurrentState)
+	}
+
+	// Unmarshal previous extra data.
+	extra := &GDSExtraData{}
+	if vasp.Extra != nil {
+		if err = vasp.Extra.UnmarshalTo(extra); err != nil {
+			return fmt.Errorf("could not deserialize previous extra: %s", err)
+		}
+	}
+
+	// Set the previous state for the new entry.
+	if len(extra.VerificationChangeLog) > 0 {
+		entry.PreviousState = extra.VerificationChangeLog[len(extra.VerificationChangeLog)-1].CurrentState
+	}
```

Comment:
This is a very smart way to determine the previous state. I am concerned that if the state is changed without an audit entry by mistake, then this will create an error - so I would want the caller to ensure the entry is correct and includes the previous state with the input. So perhaps we could make this optional? e.g.

```go
if entry.PreviousState == 0 && len(extra.VerificationChangeLog) > 0 {}
``` 

That way this will not overwrite the caller's entry.